WIP: Unity c

nullorvoid opened the pull request:
Work in progress will need review.

ronkorving commented:
Please try to explain where these rules come from. If they're based on the Unity official style guide, please refer to it. If it's based on something else, refer to that please :)

nullorvoid commented:
Did you want me to source MSDN / Unity at the top / bottom of the file?

MSDN / Unity Style guide are the same with regards to this, however because whoever wrote the Unity Style guide didn't use Unity (or got to the hard part in 2013 and stopped) we (T&J) found that the use of public / private styles should be done differently.

ronkorving commented:
So perhaps just summarize that (or go virtually verbatim) in a few lines near the top of the readme?

nullorvoid commented:
Ok will do!  Will stick a notes esq section at the top with our use cases.  This is still open for debate, I did link it to @snaptothegrid in hipchat earlier to get that side of it.  If you have any problems, or things you think should be different, do poke to discuss!

ronkorving commented:
I really don't care about *what* it becomes, as long as the reasoning behind it is solid. That's why I asked for those references basically. They at least partly explain the origins and the reasoning.

nullorvoid commented:
updated with preamble, I tried to link to the correct areas of msdn.

ronkorving commented:
Thanks :)

ronkorving commented on Unity-C#/README.md:
```diff
@@ -0,0 +1,254 @@
+
+MSDN Class Styles documenting that public fields are not supposed to be used: http://msdn.microsoft.com/en-us/library/ms229012(v=vs.110).aspx
+
+MSND XML Comments: http://msdn.microsoft.com/en-us/library/b2s063f7.aspx
```
MSND -> MSDN

nullorvoid commented:
Updated with corrections, not sure if I need the second for inher / interfaces for C#.

MiLk commented on Unity-C#/README.md:
````diff
@@ -0,0 +1,252 @@
+
+## Object and Array creation
+```csharp
+
````
Nothing here?

nullorvoid replied:
Not sure its required, I was thinking about putting in array instantiation and the use of var key word, but I'm not sure it is necessary.

MiLk commented on Unity-C#/README.md:
```diff
@@ -0,0 +1,252 @@
+### Comment Syntax
+The Visual Studio syntax is based on XML . Many tools can extract this XML and create useful help documents for you. These comments must be well-formed.
+
+It is important to note that projects in Unity may not be using Visual Studio, in this case one style should be picked, and be accepted project wide.
```
In the case where the VS syntax is not used, instead of letting people using a syntax which is unusable, why not force something else like doxygen? which can be used to generate proper documentation.

nullorvoid replied:
Looks awesomely complex ... Not sure we need to document more than one, maybe we should add a clause for using one on a project, although I would prefer to stick with just either VS as its super simple, or basic comments.

MiLk commented:
Looks like most of the `SHOULD` should be `MUST`.

ronkorving commented on Unity-C#/README.md:
```diff
@@ -0,0 +1,252 @@
+
+## Naming
+All names SHOULD be written in English, American English.
+### In general:
```
This is the time to talk about case styles.

nullorvoid replied:
.. not sure what you are looking for here with this comment.  Can you go a bit deeper in what you want here?

snaptothegrid replied:
How a bout general functions, is uppercase naming advised?

nullorvoid replied:
yes functions are upper case

ronkorving commented on Unity-C#/README.md:
````diff
@@ -0,0 +1,252 @@
+```
+
+## Return statement
+To avoid deep nesting of if-statements, always return a function's value as early as possible.
````
This slightly conflicts with line 124.

nullorvoid replied:
Not really, I think the one in 124 means that you should rank your if / else in terms of priority, and this statement is talking about returning the function as early as possible rather than having if - if else - if else - else - return val

ronkorving commented on Unity-C#/README.md:
````diff
@@ -0,0 +1,252 @@
+## Conditions
+Any non-trivial conditions should be assigned to a descriptive variable
+```csharp
+bool isAuthorized = (user.IsAdmin() || user.IsModerator());
````
While we're talking about style, *personally* I'm really not a fan of putting braces around `user.IsAdmin() || user.IsModerator()` like you do here :)

nullorvoid replied:
Probably not for two variables but if we had some other complex logic then I think it is necessary, so for consistency I would put them in. Though I'm open for suggestions to remove them.  Was there any technical reason that it was used in the JS version?

ronkorving commented on Unity-C#/README.md:
```diff
@@ -0,0 +1,252 @@
+The use of "break" and "continue" is not discouraged.
+
+### Conditionals
+Complex conditional expressions SHOULD be avoided; use temporary boolean variables instead with proper naming.
```
Is this not a duplication of what you describe on line 139?

nullorvoid replied:
It is, @micky2be did you want me to PR the JS version with some of these comments also, or have you started working on this already?

ronkorving commented on Unity-C#/README.md:
```diff
@@ -0,0 +1,252 @@
+
+
+### Comment Syntax
+The Visual Studio syntax is based on XML . Many tools can extract this XML and create useful help documents for you. These comments must be well-formed.
```
Will XML never die? :(

nullorvoid replied:
No never!

ronkorving commented on Unity-C#/README.md:
```diff
@@ -0,0 +1,252 @@
+/// <summary>
+/// Renames the user
+/// </summary>
+/// <param name="name">Persons name</param>
```
The old name or the new name? Also "Person's" or "The person's". You also mix "user" and "person".

Suggestion for this param description: "The new name to assign to the user"

nullorvoid replied:
will fix.

ronkorving commented on Unity-C#/README.md:
````diff
@@ -0,0 +1,252 @@
+Method descriptions should start with a sentence written in the third person declarative voice.
+```csharp
+/// <summary>
+/// Renames the user
````
To "Bob"? Or to something I pass? :) The summary here just replicates the name of the method, which seems rather useless.

Suggestion: "Assigns the provided name to the user".

nullorvoid replied:
will fix.

ronkorving commented on Unity-C#/README.md:
```diff
@@ -0,0 +1,250 @@
+
+
+## Naming
+All names SHOULD be written in English, American English.
```
Ironic that this document is written in British English ;) (I don't really care about that btw)

micky2be replied:
I rewrote this sentence on my side btw. 
Now it reads `All names SHOULD be written in American English.`
:stuck_out_tongue_winking_eye:

daniel-dressler replied:
So why American English? As a Canadian I work by default using either but if I understand the company's make up most programmers here would default to Commonwealth English right? Or does the French school system use American? Or maybe is it in attempt to align with how the majority of computer science culture is american focused?

nullorvoid replied:
Its because code is American "Color" rather than "Colour" for instance, I believe Wizcorp have been bit by this before in an application.

daniel-dressler replied:
Good reasoning. I am satisfied.

ronkorving replied:
Indeed. CSS for example uses "color", as I think every programming environment out there does. The day Ally decided (out of some weird sense of pride) to name a property "colour", people were forced to remember in which cases in their code base to use one or the other. It's mental load (confusion) that helps absolutely nobody.

nullorvoid commented:
Updated with Events and Delegates,  I've left them as upper case, I did this because I feel that they are closer to Methods than properties that should be viewed in the Unity editor.

/opinions?

MiLk commented:
LGTM

nullorvoid commented:
Comments from closed PR for reference:

1. Concerns about the comments ... added links to references, suggestions for more alternatives welcome.
2. Level depth for some sections ... corrected
3. Variable names sentence ... need to address, seems fine to me
4. Const ... converted to full upper case

daniel-dressler commented on Unity-C#/README.md:
````diff
@@ -0,0 +1,263 @@
+i += 1;
+```
+
+Try to put comment at the top of a block of code to explain the point of it..
````
Is this line addressing the issue of commenting functions?

Me and Rob had a debate last night about the prevalence of boiler plate comments. My argument was that the process requirement of commenting all functions with a XML comment led to the misleading and outdated comments. To summarize, perhaps unfairly, Rob argued that instead it was time constraints which caused the comments to become of negative value.

So let me open the floor by suggesting that "Comments should be treated like code. Just as we would not let dead code through review neither should we allow trite or outdated comments. If nothing meaningful can be written in a comment then do not write a comment."

So the point of contention would be if comments should be special or common.

I would frame it to say that special comments are those which warn, or explain things which deserve note. As established prior in the style guide bad code should be refactored. Thus comments represent a failure state. They are special because they mark unusual code. Upon my argument the hypothetical world's best codebase would have no comments. All function names would be succulent and clear. No abstractions would leak. All IDE highlighting would be auto generated from the code not by programmers.

Upon my argument documentation must exist outside the codebase. My favourite documentation is examples and turorials. If I have come so far into a new project or framework that I am reading code then I will read the code. A good example of what I consider good documentation is the Three.js project. They have many examples and their documentation does not clutter the codebase.

My least favourite documentation is the Bullet Physic library's. Their documentation is auto generated and trite. The codebase is filled with meaningless comments cluttering the code. The documentation generated from these comments does not explain anything. Example: http://www.continuousphysics.com/Bullet/BulletFull/classbtCollisionWorld.html

The collisionWorld is a super important vital must-use interface. Yet the documentation does not tell you how to use it.

So, should comments be special or common?

ronkorving replied:
I absolutely have to agree with you on most points you make here, @daniel-dressler . Comments are not supposed to rewrite the code in human-readable-language. We're programmers. Code is already supposed to be human readable. That's the whole point of a style guide: ensure that the code is self-explanatory. The utopian code base indeed carries no comments (but I don't think we'll live to see this happen). Variable names, argument names, function names, everything should describe its own reason to exist and what it does through the virtue of a well chosen name.

So when should we comment? When the language won't let you do all you need to communicate some of what I wrote above (when strong typing is lacking for example, because a type does communicate). That's why we use JSDoc in JavaScript. Comments are also useful when you're about to write a function that needs to do something complicated. You don't repeat the algorithm (it will get out of sync with reality). You describe what it does (not how it does it, the code *should* suffice). Also, I believe one should *never* repeat values in comments (eg: `int length = 5; // length is 5`), because the moment they're out of sync, you have nothing to fall back to in order to discover which is true (is this a bug, or just a bad comment?).

There are *always* going to be exceptions to these "rules". Performance sensitive code sometimes just ends up being ugly. I'm sure OS kernels and game engines are full of it. That's when you comment the crap out of the solution to make it really, really, really clear what that ugly piece of code achieves and how it does that. And you apologize to the next maintainer :)

TL;DR: Ideally, comments describe what and code describes how. They should not tell the same story twice.

nullorvoid replied:
Recently I was thinking how nice C++ has it with functions declarations where you can have a page of comments and then the actual implementation separate.  eg..

/** Removes from it's parent, including the viewport if it exists */
virtual void RemoveFromParent() override;

/**
 * Sets the widgets position in the viewport.
*/
void SetPositionInViewport(Vector3 Position);

/*  */
void SetSizeInViewport(Vector3 Size);


I'm actually very against the coders dream of no comments self describing code, because you'll end up with things like in .Net

using System.Windows.Media.TextFormatting.TextSource;
GetTextEffectCharacterIndexFromTextSourceCharacterIndex();

Although I agree we should be pragmatic, I want to answer this with proper responses and examples of where we can gain automated benefits from the XML comments, so I will try to write an example for here later in the week / next week.

ronkorving replied:
I think the problem you describe there has more to do with architectural bloat than anything else. Java also suffers from that. I don't really see comments solving that (how would they, the comments sit with the implementation?), apart from of course the tight editor integration you're probably thinking of. I'm arguing for human understandable naming conventions. Your example, I don't even understand what that does (like, at all) :)

nullorvoid replied:
XD I have no idea either, but i think it would be better to have 
/// <Summary>
/// Some really useful comment that actually describes what this function is all about.
/// </Summary>
GetTextEffectCharacterIndex()
{
// Blah
}

ronkorving replied:
I feel I have to make one thing really clear. I'm not against functions having comments above them (on the contrary!). But I do think the comment should not describe how it achieves what it achieves. It should describe what it does (not how), and maybe why it needs to exist in the first place (use cases).

daniel-dressler commented on Unity-C#/README.md:
```diff
@@ -0,0 +1,263 @@
+
+
+## Braces
+In C# where the braces are placed is up to the user, however the C# MSDN standard is on a new line.  That said as long as the project has agreed to use on the same line there should be no complaints.
```
For the most part this C# code style matches our JS style. The new line brace style is the only major departure that I know of. May I suggest we break from MSDN and settle with continuing the JS brace style?

Justification: Consistency with our JS style would give us low overhead when switching languages. In theory within the Almir team we would expect frequent switching as programmers are not dedicated to backend vs frontend roles. 

So, would we expect any costs from breaking MSDN compatibility? What are we gaining from the new line braces? Is consolidation between languages worth losing these benefits?

nullorvoid replied:
I think there are for's and against's, I have sourced most of these and came to the conclusion that Unity developers were not thinking of styles of the other languages when they wrote the engine.

I would read the Unity style guide (which basically follows the MSDN guidelines)

I would work with the engine and have a look at the styles (or lack there of) from the engine code itself.

I have considered JS developers as it is OK to use one or the other, but it should depend on project, one project shouldn't have both (TJ IS THE WORST EXAMPLE OF THIS)

I have considered what C# is and where is came from rather than just JS developers.  For new hires that are Unity developers and not JS developers the C# standard is MSDN (for the most part) because C# is Microsofts language, so it makes sense when you start C# you tend to start with MSDN.

The trade off is against the C# standard / Unity standard, or JS standard.  Each is a different language, JS has the brackets on the same line because it HAS to, C# this restriction is not there.

ronkorving replied:
I'm in favor of sticking to C# and Unity standards. If you're going to be handling these tools and the various libraries and plugins that exist in the ecosystem, it's *that* ecosystem you want to feel comfortable around. We can't go and try to change every language into JavaScript, just because that's what we used mostly today. We need to invest our energy in being the best C# and Unity developers we can be (not JavaScript users that accidentally found themselves in the "wrong" project), and part of that is our ability to interact with online communities. If every time you open someone else's C# code you think "yuch, curly brace horror!", you're exactly on the wrong path imho.

daniel-dressler commented on Unity-C#/README.md:
```diff
@@ -0,0 +1,263 @@
+#Preamble 
+The below style uses the MSDN Style guidelines as a base to work with Unity, however because some of the nuances in Unity this style adapts and breaks from some of the standard conventions to obtain a higher level of readability.
```
From my reading this document can stand on its own. I feel the MSDN reference should come at the document's bottom and be treated as "Anything not defined here should defer to MSDN".

On the whole I think the frequent references to MSDN and justifications for departures from MSDN style should be in a separate section along with the references.

I think the document would also benefit from a section targeted at JS programmings. This section would explain what is different and call out features of C# which require special attention.

Thus in concrete terms I think this document should not depend on MSDN. Most sections should explain the WHAT and the WHY but drop the WHY NOT MSDN. The WHY NOT MSDN could be useful in a dedicated section.

nullorvoid replied:
Good suggestion, I also agree, I put it at the top so it was clear as per a different suggestion in this PR.  We can restructure some of this definitely.

daniel-dressler commented on Unity-C#/README.md:
```diff
@@ -0,0 +1,263 @@
+/// <summary>
+/// Assigns the provided name to the user
+/// </summary>
+/// <param name="name">The new name to assign to the user</param>
```
Would this be the proper place to point out that this documentation does not appear to justify itself?

The pareter name should likely be something like "newName", if nothing else I would expect this Class to already contain a "name" field. With this alteration the comment does not appear to provide anything a glance at the function declaration would not provide.

Of course we charge by the man month so there could be a business case for busy work.

nullorvoid replied:
Yes and no .. the point of this simple example is to show "Method descriptions should start with a sentence written in the third person declarative voice." which it does and is a simple example of this, the method itself and what it does is completely irrelevant to the point of this, although both you and Ron have picked up on this, so I think it should be rewritten to be a better method, I can change this no problem.
